docs/design/detection_of_genenated_infinities_and_nans.md

```diff
@@ -0,0 +1,294 @@
+  fd = fopen(...);
+  if (!cond) {
+    fclose(fd);
+  }
+  fprintf(...); // (1)
+  if (cond) {
+    fclose(fd);
+  }
+
+  // FALSE NEGATIVE: The file descriptor opened at (2) flows into the dispose
+  // call at (3) if the values of x and y are not known. However, the resource
+  // is only closed when x == y, which is not necessarily the case.
+  fds[x] = fopen(...); // (2)
+  fclose(fds[y]);      // (3)
+```
+
+Nevertheless, their approach is sensible and likely good enough.
+
+Lastly, this approach has the unfortunate downside that unused float values which could be NaN or Infinity will be reported, when they do not have any negative effect on a program (as opposed to the negative effects of leaking unused file descriptors, or unused memory, etc).
+
+## Case study examples
+
+The following is an interesting set of examples and code snippets that come from the open source project [pandas](https://github.com/commaai/panda), which aims to be MISRA compliant and is used for self-driving cars.
+
+These examples are hand picked results from a query that selected expressions with a floating point type along with their upper and lower bounds.
+
+**Example 1**:
+```c
+float filtered_pcm_speed =
+  ((to_push->data[6] << 8) | to_push->data[7])
+  * 0.01 / 3.6;
+// Disable controls if speeds from ABS and PCM ECUs are too far apart.
+bool is_invalid_speed = ABS(filtered_pcm_speed
+  - ((float)vehicle_speed.values[0] / VEHICLE_SPEED_FACTOR))
+  > FORD_MAX_SPEED_DELTA;
+```
+
+While `filter_pcm_speed` cannot be infinity or NaN, it is interesting to see how this value is sanity checked. If this code were refactored such that it could produce NaN, the greater-than check would return false, resulting in a bug. If the condition were flipped (check inside valid range, rather than outside), it would handle NaN correctly. This cannot be captured via static analysis.
+
+**Example 2**:
+```c
+        float x0 = xy.x[i];
+        float y0 = xy.y[i];
+        float dx = xy.x[i+1] - x0;
+        float dy = xy.y[i+1] - y0;
+        // dx should not be zero as xy.x is supposed to be monotonic
+        dx = MAX(dx, 0.0001);
+        ret = (dy * (x - x0) / dx) + y0;
+```
+
+This is an [interpolation function](https://github.com/commaai/panda/blob/dec9223f9726e400e4a4eb91ca19fffcd745f97a/board/safety.h#L538), where `xy` is a struct parameter, with array members `x` and `y` that represent points in the domain and range to interpolate across.
+
+Range analysis is performed with local information only, and therefore, the expression `xy.x[i]` is given the range [-Infinity, Infinity]. This is not a generated infinity. However, the computations of `dx` and `dy` could generate a positive or negative infinity (if both numbers are finite and the result exceeds the maximum float value), they could propagate a positive or negative infinity, and/or they could generate a NaN (if an infinite value is subtracted from itself).
```

Comment:
Hey Mike, about 

> the computations of `dx` and `dy` could generate a positive or negative infinity (if both numbers are finite and the result exceeds the maximum float value),

I tried to reproduce (https://godbolt.org/z/8sTTs6Mc9). Is it C17 standard that addition or subtraction of finite floating point numbers can lead to infinity?